// node-red-contrib-noble/trainer/xpd-trainer.js
/*
 * Copyright (c) 2014. Knowledge Media Institute - The Open University
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * NodeRed node with support for interaction with BLEs
 *
 * @author <a href="mailto:carlos.pedrinaci@open.ac.uk">Carlos Pedrinaci</a> (KMi - The Open University)
 * based on the initial node by Charalampos Doukas http://blog.buildinginternetofthings.com/2013/10/12/using-node-red-to-scan-for-ble-devices/
 */
module.exports = function(RED) {
    
    
    // The main node definition - most things happen in here
    function XpdTrainerNode(n) {
        // Create a RED node
        RED.nodes.createNode(this,n);

        // Store local copies of the node configuration (as defined in the .html)
        this.data = n.data;
        this.name  = n.name;
        this.category = n.category;

        var node = this;



        node.on('input', function (msg) {

            if (node.data && node.category) {
                var json = JSON.parse(node.data);
                is_one = true;
                for(var k in json){
                    if (typeof(json[k]) == 'object'){
                        is_one = false;
                        msg["input"] = json[k];
                        var outObj = {};
                        outObj[node.category] = 1;
                        msg["output"] = outObj;
                        node.send(msg);
                    }
                    else{
                        is_one = true;
                    }
                }

                if(is_one){
                    console.log(json);
                    msg["input"] = json;
                    var outObj = {};
                    outObj[node.category] = 1;
                    msg["output"] = outObj;
                    node.send(msg);
                }

                
                
            }
        });

    
        node.on("close", function() {

        });

        function toArr(obj){
            var re={};
            (function f(o){
                for(var k in o)
                    if(typeof o[k]=='object'){
                       var k2=k.replace(/['"]/g,'');
                       /^\d+$/.test(k2)&&(re[k2]=o[k]),f(o[k]); 
                    }
            })(obj);
            return re;
        }
    }


    
    // Register the node by name. This must be called before overriding any of the
    // Node functions.
    RED.nodes.registerType("xpd-trainer",XpdTrainerNode);

}
